### DISCUSSION TITLE
iOS Support

### DISCUSSION THREAD

**Discussion by abbanatic** [jamulussoftware/jamulus#1259] [Ideas]
Hi,

please be so kind to think about providing iOS support for Jamulus in the future.

Due to Apple's change from Intel to ARM, universal binaries for all of Apple's platforms became reality.

So it would be great for Jamulus to support the New Apple Silicon AND iOS devices, with their great built-in capabilities regarding Audio and MIDI.

Thank you very much!
Thomas

**Comment by ann0see** [comment]
**EDIT 2026:** There's an [official iOS release on the AppStore](https://apps.apple.com/it/app/jamulus/id6654925705).

**EDIT**: Thanks to multiple people there is basic iOS and Android Support for Jamulus. Please see the https://github.com/jamulussoftware/jamulus/blob/master/COMPILING.md#ios file for information

***

Agree. Porting should be possible but there would still be the same two problems like for Android (which is in development):

1. how to deal with WiFi (of course you could use USB to LAN adapters)
2. How to deal with Audio Interfaces (iOS uses CoreAudio, so does macOS)

Something iOS specific:

Submitting to the AppStore would require Jamulus to be dual licensed (?). Apple doesn't like the GPL. 

See https://www.reddit.com/r/linux/comments/fv1hmp/discuss_those_scenarios_for_gpl_and_apple_app/

And the Nextcloud iOS Repo on GitHub.

See also #83 Android Support

**Comment by ann0see** [comment]
First of all, I don't have any knowledge of iOS development. Therefore somebody experienced should look at this issue. 

I want to document some changes I made compile and run Jamulus on iOS:

1. I created a new iOS folder (with the content of the macOS files), added the Info.plist file and edited the .pro file.

I tried to compile Jamulus with QT Creator for iOS and it (of course) failed. 
We can't use the same audio backend as in macOS:
`In file included from /.../jamulus/ios/sound.cpp:25:
/.../jamulus/ios/sound.h:70:5: error: unknown type name 'AudioDeviceID'; did you mean 'AudioFileID'?` 

This will probably be one of the main issues. We will need to rewrite this part for iOS.

Changes needed to compile for iOS:

1. In client.h the include statement for Apple systems has to be changed to check if it's being built for iOS.

2. Also clientdlg.cpp: SetMyWindowTitle: we can't show a badge number on iOS --> exclude it for iOS

1. Framework AudioUnit doesn't exist on iOS --> don't include it in the .pro file


I could now at least compile it via XCode. But it only shows a black screen. I don't know how to get iOS show the QT widgets. Maybe somebody from the Android team knows something about that? Do we need a complete UI rewrite?

If anybody has experience I can try and test it.

**Comment by gene96817** [comment]
Apple announced at WDC2020 some kind of merging macOS and iOS for 2021. I don't follow the details in this area closely. This suggests people interested in iOS support should look at the developer's tools for macOS 12.x.x and this "merger" with iOS. Of course, this doesn't solve the needs of devices that run old iOS release.

**Comment by ann0see** [comment]
I got a headless Jamulus server up and running on my iPhone SE 1st gen (although it's connected via wifi, the quality is ok) which can show up on a central server.

Everything GUI related doesn't work and also I assume files can't be written because of the iOS sandbox.

**Comment by ann0see** [comment]
https://developer.apple.com/documentation/avfoundation/avaudiosession?language=objc in https://developer.apple.com/documentation/avfoundation/avaudiosessionmodemeasurement mode could be a good choice for audio in iOS. Does anybody know anything about iOS Audio/iOS CoreAudio?

**Comment by ann0see** [comment]
Although I have no idea how to write the sound.cpp or sound.h files or StoryBoard files etc. have a look at my changes: 

https://github.com/ann0see/jamulus/tree/patch-3

**Comment by ann0see** [comment]
Anybody else here knowing iOS development and QT? Unfortunately I have almost no idea how the csound class should look like on iOS since it seems as if we can't use the same code line on macOS. 

@corrados what are the main methods for a sound.h/sound.cpp file? 

Did I find the correct ones in my code?

**Comment by corrados** [comment]
Is this still work in progress? Or shall I remove the label?

**Comment by ann0see** [comment]
You should remove the label (but I will probably continue working on some of this stuff once the installer and autobuild stuff works)

**Comment by ann0see** [comment]
Current open issues here:

1. Write sound driver for iOS
2. Fix the GUI: Currently if you launch Jamulus on iOS (after having it compiled via Xcode), it displays the "Would like to access your microphone" screen but doesn't show anything else. I think the problem is related to some settings in Xcode I don't know. There's almost no documentation for Qt widgets applications on iOS and I don't think a Qt widgets app would ever get into the AppStore.

**Comment by ann0see** [comment]
See this PR: https://github.com/jamulussoftware/jamulus/pull/1512 sound and GUI now work on some devices. It's still not yet finished

**Comment by paul-cine** [comment]
This really need to happen. Have you considered reaching out to some of the best audio devs on iOS.
They might be able to assist.
First I would try the devs of Loopy Pro and AUM.
https://loopypro.com/
http://kymatica.com/apps/aum
Or ask for assistance on the audiobus forum.

**Reply by ann0see** [reply]
Thanks for the info. We do have iOS working but currently it breaks visually on Qt6.

**Reply by ann0see** [reply]
This means audio does work but the UI needs work.

**Comment by admin** [comment]
Jamulus will soon be available on the AppStore: https://apps.apple.com/us/app/jamulus/id6654925705
